maskit/masks.py

```diff
@@ -0,0 +1,242 @@
+    ):
+        """
+        Perturbs the MaskedObject by the given ``mode`` of type
+        :py:class:`~.PerturbationMode` ``amount`` times. If no amount is given,
+        that is ``amount=None``, a random ``amount`` is determined given by the
+        actual size of the py:attr:`~.mask`. The ``amount`` is automatically
+        limited to the actual size of the py:attr:`~.mask`.
+
+        :param amount: Number of items to perturb, defaults to None
+        :param mode: How to perturb, defaults to PerturbationMode.INVERT
+        :raises NotImplementedError: Raised in case of an unknown mode
+        """
+        assert (
+            amount is None or amount >= 0
+        ), "Negative values are not supported, please use PerturbationMode.REMOVE"
+        if amount == 0:
+            return
+        count = abs(amount) if amount is not None else rand.randrange(0, self.mask.size)
+        if mode == PerturbationMode.ADD:
+            indices = np.argwhere(~self.mask)
+        elif mode == PerturbationMode.INVERT:
+            indices = np.array([list(index) for index in np.ndindex(*self.mask.shape)])
+        elif mode == PerturbationMode.REMOVE:
+            indices = np.argwhere(self.mask)
+        else:
+            raise NotImplementedError(f"The perturbation mode {mode} is not supported")
+        if len(indices) == 0:
+            return
+        indices = tuple(
+            zip(
+                *indices[
+                    np.random.choice(
+                        len(indices), min(count, len(indices)), replace=False
+                    )
+                ]
+            )
+        )
+        self.mask[indices] = ~self.mask[indices]
+
+    def copy(self) -> "Mask":
+        """Returns a copy of the current MaskedObject."""
+        clone = object.__new__(type(self))
+        clone.mask = self.mask.copy()
+        return clone
+
+
+class MaskedCircuit(object):
+    """
+    A MaskedCircuit supports masking of different components including wires,
+    layers, and parameters.
+    """
+
+    __slots__ = (
+        "_layer_mask",
+        "_wire_mask",
+        "_parameter_mask",
+        "parameters",
+    )
+
+    def __init__(self, parameters: np.ndarray, layers: int, wires: int):
+        assert (
+            layers == parameters.shape[0]
+        ), "First dimension of parameters shape must be equal to number of layers"
+        assert (
+            wires == parameters.shape[1]
+        ), "Second dimension of parameters shape must be equal to number of wires"
+        self.parameters = parameters
+        self._parameter_mask = Mask(shape=parameters.shape)
+        self._layer_mask = Mask(shape=(layers,))
+        self._wire_mask = Mask(shape=(wires,))
+
+    @property
+    def mask(self) -> np.ndarray:
+        """
+        Accumulated mask of layer, wire, and parameter masks.
+        Note that this mask is readonly.
+        """
+        mask = self.parameter_mask.copy()
+        mask[self.layer_mask] = True
```

Comment:
As discussed, *I think* it would be more readable to do this:
```suggestion
        mask[self.layer_mask,:] = True
```
But I have no idea whether that is correct. ;)

Reply:
Looks reasonable... Tests also work. Let's give it a try 👍